### PR TITLE
Look for orig$ functions from side modules too

### LINKED ISSUE #9901: Function Pointers and dynamic linking (function signature mismatch)
I'm trying to use boost regex expressions with a custom build icu (and latest tot upstream emscripten). I always get an error
```
exception thrown: RuntimeError: function signature mismatch,RuntimeError: function signature mismatch
    at legalstub$dynCall_jiii (wasm-function[12507]:11)
    at Object.Module.dynCall_jiii (/Test.js:19640:40)
    at invoke_jiii (/Test.js:27738:34)
    at legalfunc$invoke_jiii (wasm-function[12528]:9)
    at boost::re_detail_106900::basic_regex_creator<int, boost::icu_regex_traits>::basic_regex_creator(boost::re_detail_106900::regex_data<int, boost::icu_regex_traits>*) (wasm-function[7314]:136)
....
```

In trying to reproduce above problem in a simpler setting I found that when using function pointers to functions in side modules that return `int64_t` values I get the same error message
```RuntimeError: function signature mismatch```. The failure seems to be similar (but still different) to #9850 which was fixed in PR #9855.

I created a small combined test case with 4 files (without the boost/icu clutter) that reproduce the problem with the function pointers:
`side.hpp`:
```c++
#ifdef __EMSCRIPTEN__
    #include <emscripten.h>
#else
    #define EMSCRIPTEN_KEEPALIVE
#endif
#include <cstdint>

EMSCRIPTEN_KEEPALIVE int32_t function_ret_32(int32_t i, int32_t j, int32_t k);
EMSCRIPTEN_KEEPALIVE int64_t function_ret_64(int32_t i, int32_t j, int32_t k);
```

`side.cpp`:
```c++
#include "side.hpp"
int32_t function_ret_32(int32_t i, int32_t j, int32_t k) {
    return 32;
}
int64_t function_ret_64(int32_t i, int32_t j, int32_t k) {
    return 64;
}
```

`main.cpp`:
```c++
#include <cstdio>
#include <cinttypes>
#include "side.hpp"

typedef int32_t (*fp_type_32)(int32_t, int32_t, int32_t);
typedef int64_t (*fp_type_64)(int32_t, int32_t, int32_t);

int32_t internal_function_ret_32(int32_t i, int32_t j, int32_t k) {
    return 32;
}
int64_t internal_function_ret_64(int32_t i, int32_t j, int32_t k) {
    return 64;
}

int main() {
    fp_type_32 fp32_internal = &internal_function_ret_32;
    fp_type_32 fp32_external = &function_ret_32;
    fp_type_64 fp64_external = &function_ret_64;
    fp_type_64 fp64_internal = &internal_function_ret_64;
    int32_t ires32 = fp32_internal(0,0,0);
    printf("res32 - internal %d\n",ires32);
    int32_t eres32 = fp32_external(0,0,0);
    printf("res32 - external %d\n",eres32);

    int64_t ires64 = fp64_internal(0,0,0);
    printf("res64 - internal %" PRId64 "\n",ires64);
    int64_t eres64 = fp64_external(0,0,0);
    printf("res64 - external %" PRId64 "\n",eres64);
    return 0;
}

```

`preload.js`:
```js
if (typeof(Module) === "undefined") Module = {};
Module.dynamicLibraries = ['side.wasm'];

```

Compiling the side module with
`em++ -g --profiling -Wall -fPIC -s WASM=1 side.cpp -s SIDE_MODULE=2  -o side.wasm`

Compiling the main module with
`em++ -o main.js -g --profiling -Wall main.cpp -s MAIN_MODULE=1 -s EXPORTED_FUNCTIONS=['_main']  --pre-js preload.js
`

Running the program leads to the RuntimeError when fp64_external is called. Without the fix  #9855 the line that should produce output `res64 - internal 64` fails.

btw: Thanks for the great work! It is a lot of fun working with emscripten.

### PR DESCRIPTION
**Pull request by kripken**
Fixes #9901

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -8813,6 +8813,65 @@ def test_main_module_without_main(self):
...
+    printf("res64 - external %" PRId64 "\n",eres64);
+    return 0;
+}
+''')
+    run_process([PYTHON, EMCC, 'side.cpp', '-s', 'SIDE_MODULE=2', '-o', 'side.wasm'])
+    run_process([PYTHON, EMCC, 'main.cpp', '-s', 'MAIN_MODULE=2', '--pre-js', 'pre.js'])
+    self.assertContained('''res32 - internal 32
+res32 - external 32
+res64 - internal 64
+res64 - external 64''', run_js('a.out.js'))
```
Comment (sbc100) [inline tests/test_other.py]: Can you not use the same helper function that other `test_dylink_*` function in this file use?

Can we reduce this test to something smaller that just tests for the exact problem?  Why both to call the functions internally for example?

File: tests/test_other.py [hunk 2]
```diff
@@ -8813,6 +8813,65 @@ def test_main_module_without_main(self):
     run_process([PYTHON, EMCC, 'src.c', '--pre-js', 'pre.js', '-s', 'MAIN_MODULE=2'])
     self.assertContained('bar', run_js('a.out.js'))
 
+  def test_dynamicLibraries_i64(self):
+    create_test_file('pre.js', r'''
+if (typeof(Module) === "undefined") Module = {};
+Module.dynamicLibraries = ['side.wasm'];
+''')
```
Comment (sbc100) [inline tests/test_other.py]: Can you just do `self.set_setting('RUNTIME_LINKED_LIBS', ['wide.wasm'])` instead of this?

File: emscripten.py [hunk 3]
```diff
@@ -1806,10 +1806,11 @@ def create_fp_accessors(metadata):
     accessors.append('''
 Module['%(full)s'] = function() {
   %(assert)s
-  // Use the wasm function itself, for the table.
+  // Use the original wasm function itself, for the table, from the main module.
   var func = Module['asm']['%(original)s'];
-  // If there is no wasm function, this may be a JS library function or
-  // something from another module.
+  // Try an original version from a side module.
+  if (!func) func = Module['_%(original)s'];
```
Comment (sbc100) [inline emscripten.py]: But why do side module symbols have an `_`?  Oh I guess they don't but when we add them to "Module" we artificially at one maybe?   Maybe we should add them to 'asm' too without the "_"?   since they are "native code symbols" too?

File: tests/test_other.py (hunk 1 above)
Comment (kripken) [reply tests/test_other.py]: I actually thought it might be good to write a new test here, as it uses `Module.dynamicLibraries`, and we have no coverage for that aside from a few in the `browser` suite. So it works differently than the existing dylink tests in `core`, and can't use the same infra.

File: tests/test_other.py (hunk 2 above)
Comment (kripken) [reply tests/test_other.py]: That only works in `core`, I believe? (We call emcc directly here.)

File: emscripten.py (hunk 3 above)
Comment (kripken) [reply emscripten.py]: Yes, I think when we add stuff to module, the support.js code adds a `_` to everything.

I don't think we can add them to `asm` - asm is an exports object from a wasm file, it's not a normal JS object.

I do think this asymmetry is not great, though - we should probably not use `asm` here, and have the main module write its symbols to `Module` like side modules do. But as the usage of `asm` was pre-existing here, I wasn't sure it makes sense to try to do a large refactoring here. But maybe you know that code better and have an idea?

File: tests/test_other.py (hunk 1 above)
Comment (sbc100) [reply tests/test_other.py]: How does that work differently?  I thought RUNTIME_LINKED_LIBS just ends up setting dynamicLibraries, no?

If you want to write a specific test for modifying dynamicLibraries I don't see that as related to fixing this i64 issue.

File: tests/test_other.py (hunk 2 above)
Comment (sbc100) [reply tests/test_other.py]: Ah, well I guess that begs the question why don't we have common patter for dylink testing in "other"?

For now, why not just add a quick, minimal dylink-style test to 'core'?

If you think that this doesn't belong in core, the same can probably be said for a lot of the other "dylink" tests too and we can do a followup change to migrate them out of core.   I think this test might actually belong in core because the JS interaction stuff is effected by things like optimization levels.

Basically I think there is way too much boiler plate in this test which it seems to me should just test for i64 function being exported from a side module.  But maybe I'm missing something?

Comment (kripken) [reply tests/test_other.py]: Not directly, you're right, this was killing two birds with one stone :smile: 

Note that I actually tried to write a quick, minimal test - and failed. After 15 minutes I felt it wasn't a good use of time, and used the existing testcase, again, given I had realized I wanted to write a new testcase for `Module.dynamicLibraries` anyhow which it uses.

I'm not sure why I failed, as the function pointer issue should be separable from how the library is loaded - but maybe that assumption is wrong? When we dlopen vs `Module.dynamicLibraries` do we add the side module symbols differently to main somehow?

Comment (sbc100) [reply tests/test_other.py]: Hmm, I'm not suggesting you use `dlopen` which I agree is a different use case.

My understanding was/is that `RUNTIME_LINKED_LIBS` and `Module.dynamicLibraries` do exactly the same thing.   I see the former as the official way to set the later.

The tests in core that use `dylink_test` use `RUNTIME_LINKED_LIBS` and not `dlopen` IIRC so they should behave exactly like adding to `Module.dynamicLibraries`.. which is to say they do library loading *before* main is called.

File: tests/test_other.py (hunk 1 above)
Comment (kripken) [reply tests/test_other.py]: I agree it's separate, but it was easier to do both at once.

If you want I can rewrite this test from scratch?

**Comment by kripken** [comment]
friendly ping @sbc100 on the questions above

**Comment by sbc100** [comment]
I still think the test should just use `RUNTIME_LINKED_LIBS`.  lgtm with that change.

We should be using a common helper for all dylib tests but that be done a followup perhaps.

File: tests/test_other.py (hunk 1 above)
Comment (kripken) [reply tests/test_other.py]: I rewrote the test in core as you preferred. It works (i.e. it fails before the PR, passes with it) - not sure what I did wrong before.
